### PR TITLE
Setup format, husky & lint-staged

### PR DESCRIPTION
**Pull request by abdonrd**
Continuing the work from #398 (from https://github.com/Polymer/lit-html/pull/1351)...

- Setup format scripts with Prettier & ESLint.
- Remove clang-format references.
- Setup husky & lint-staged.

I save the auto-format to just one commit to easy review: https://github.com/webcomponents/polyfills/pull/405/commits/e509977f60a9687b09ce7dfc4dc58edbdb50acbe

With this PR, the mono-repository has exactly the same configuration as the `lit-html` mono-repository.

### CODE REVIEW

**Comment by abdonrd** [comment]
@bicknellr do you know why the timeouts in the tests?

**Comment by abdonrd** [comment]
@bicknellr now it's all working! But do you know why these changes (https://github.com/webcomponents/polyfills/pull/405/commits/bf453a3049c26d00ca8e987e314b6c43aae0d16b) break the tests?

**Comment by abdonrd** [comment]
Oh, it looks like a Closure restriction 😕 

> https://developers.google.com/closure/compiler/docs/limitations
> Using string names to refer to object properties

**Comment by abdonrd** [comment]
@bicknellr Yay! Everything works now! 🎉

**Comment by abdonrd** [comment]
Friendly ping to @bicknellr

**Review by bicknellr — CHANGES_REQUESTED** [review]
Hey! Sorry about the big delay, I'm a bit swamped with internal tasks currently. The majority of these comments are just that the formatter seems to be adding self-closing-tag-style slashes to void elements. There are also a couple of things w.r.t. closure minification as well as the formatter removing some of the `;`s from the end of mixin blocks in CSS, which might just require telling the linter to ignore those sections since it's invalid syntax. Also, can we exclude the license files? I'm actually not sure why those are markdown rather than plain text, but I'd rather not have anything automatically editing them.

File: packages/html-imports/README.md [hunk 1]
```diff
@@ -54,15 +54,19 @@ In IE/Edge, appending `<link rel=stylesheet>` in a node that is not `<head>` bre
 `my-element.html` imports a stylesheet and applies a style:
 
 ```html
-<link rel="stylesheet" href="my-linked-style.css">
-<style> .blue { color: blue }; </style>
+<link rel="stylesheet" href="my-linked-style.css" />
```
[the remark this answers is not in the document]
Comment (abdonrd) [reply packages/html-imports/README.md:57]: We can't. Prettier is a very opinionated code formatter; has few config options.

File: packages/shadydom/src/patch-events.js [hunk 2]
```diff
@@ -56,71 +59,71 @@ const parseEventOptions = (optionsOrCapture) => {
...
+  DOMAttrModified: true,
+  DOMAttributeNameChanged: true,
+  DOMCharacterDataModified: true,
+  DOMElementNameChanged: true,
+  DOMNodeInserted: true,
+  DOMNodeInsertedIntoDocument: true,
+  DOMNodeRemoved: true,
+  DOMNodeRemovedFromDocument: true,
+  DOMSubtreeModified: true,
+};
```
[the remark this answers is not in the document]
Comment (abdonrd) [reply packages/shadydom/src/patch-events.js:126]: I have no idea about closure, but we can config this with: https://prettier.io/docs/en/options.html#quote-props

**Comment by abdonrd** [comment]
@bicknellr 

- We can't avoid the self-closing-tag-style slashes. Prettier is a very opinionated code formatter; has few config options.
- Updated the quote props! Now it preserve them.

**Comment by abdonrd** [comment]
@bicknellr 

- We can't avoid the self-closing-tag-style slashes. Prettier is a very opinionated code formatter; has few config options.
- Updated the quote props! Now it preserve them.
- Manually back the `;` from the mixin blocks in CSS;

**Comment by abdonrd** [comment]
Friendly ping to @bicknellr 😊

**Comment by abdonrd** [comment]
@bicknellr rebased again to `master`!

File: packages/url/urltestgenerator.html [hunk 3]
```diff
@@ -26,11 +26,11 @@
         continue // only true for "_password"
       }
       if(output !== "")
-        output += " "
+        {output += " "}
       output += componentsKeys[i] + ":" + esc(val)
     } else if(components[i] == "_password") {
```
Comment (bicknellr) [inline packages/url/urltestgenerator.html]: Is prettier set up to run on this file? This isn't something I'd expect it to produce.

File: packages/formdata-event/ts_src/formdata_listener_added.ts [hunk 4]
```diff
@@ -167,19 +188,22 @@ export const wrapSubmitListener = (listener: EventListenerOrEventListenerObject)
     // event.
     if (getEventPropagationImmediatelyStopped(e) && targetIsAForm) {
       maybeDispatchFormdataForEvent(e);
-    // If the event's propagation was stopped (potentially before being seen by
-    // the capturing 'submit' listener) and this is the _last_ callback for its
-    // event phase, dispatch a 'formdata' event.
+      // If the event's propagation was stopped (potentially before being seen by
+      // the capturing 'submit' listener) and this is the _last_ callback for its
+      // event phase, dispatch a 'formdata' event.
     } else if (getEventPropagationStopped(e) && targetIsAForm) {
```
Comment (bicknellr) [inline packages/formdata-event/ts_src/formdata_listener_added.ts:194]:
```suggestion
    }
    // If the event's propagation was stopped (potentially before being seen by
    // the capturing 'submit' listener) and this is the _last_ callback for its
    // event phase, dispatch a 'formdata' event.
    else if (getEventPropagationStopped(e) && targetIsAForm) {
```

Maybe sticking the `}` before the comment would keep the indentation correct?

File: packages/formdata-event/ts_src/formdata_listener_added.ts [hunk 5]
```diff
@@ -167,19 +188,22 @@ export const wrapSubmitListener = (listener: EventListenerOrEventListenerObject)
...
+      ) {
         maybeDispatchFormdataForEvent(e);
       }
-    // If this listener is the _last_ bubbling 'submit' event listener attached
-    // to the shallow root of the event's target (possibly the one added by the
-    // capturing listener), dispatch a 'formdata' event.
+      // If this listener is the _last_ bubbling 'submit' event listener attached
+      // to the shallow root of the event's target (possibly the one added by the
+      // capturing listener), dispatch a 'formdata' event.
     } else if (submitEventToListenerInfo.has(e)) {
```
Comment (bicknellr) [inline packages/formdata-event/ts_src/formdata_listener_added.ts:207]:
```suggestion
    }
    // If this listener is the _last_ bubbling 'submit' event listener attached
    // to the shallow root of the event's target (possibly the one added by the
    // capturing listener), dispatch a 'formdata' event.
    else if (submitEventToListenerInfo.has(e)) {
```

(same comment as above)

**Review by bicknellr — CHANGES_REQUESTED** [review]
Thanks for keeping this up to date! Btw, you can resolve conflicts without having to move previous commits by merging master into your own branch. I think it helps GitHub point out what's changed between reviews. (very useful for mass changes like this too)

File: packages/shadydom/src/patches/Node.js [hunk 6]
```diff
@@ -488,7 +537,8 @@ export const NodePatches = utils.getOwnPropertyDescriptors({
    * @param {Object=} options
    */
   // TODO(sorvell): implement `options` e.g. `{ composed: boolean }`
-  getRootNode(options) { // eslint-disable-line @typescript-eslint/no-unused-vars
+  getRootNode(options) {
+    // eslint-disable-line @typescript-eslint/no-unused-vars
```
[the remark this answers is not in the document]
Comment (abdonrd) [reply packages/shadydom/src/patches/Node.js:543]: Removed linter comment.

About the `implement options` probably not, but maybe @sorvell can confirm it.

File: packages/tests/shadycss/css-parse.html [hunk 7]
```diff
@@ -9,180 +9,244 @@
...
+    </style>
+
+    <style id="multiple-spaces">
+      .foo .bar {
+      }
+      .foo .bar {
+      }
+      .foo .bar {
+      }
+    </style>
```
[the remark this answers is not in the document]
Comment (abdonrd) [reply packages/tests/shadycss/css-parse.html]: Done!

File: packages/url/urltestgenerator.html (hunk 3 above)
Comment (abdonrd) [reply packages/url/urltestgenerator.html]: Yes! In all files: https://github.com/webcomponents/polyfills/pull/405/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R13

**Comment by abdonrd** [comment]
@bicknellr thanks! I think everything is done!

**Comment by bicknellr** [comment]
(adjusted CLA tags because googlebot doesn't like my personal email)

**Comment by abdonrd** [comment]
> (adjusted CLA tags because googlebot doesn't like my personal email)

Thanks!

**Comment by bicknellr** [comment]
BTW, https://github.com/webcomponents/polyfills/pull/383 was merged and there are a few conflicts so you'll need to merge from master again - fortunately it's mostly clean.

I spent some time today looking into if there's a way to get prettier to ignore both files that match the globs in `.gitignore` as well as others without duplicating `.gitignore`, but it seems like this isn't possible right now. (https://github.com/prettier/prettier/issues/8048, https://github.com/prettier/prettier/issues/8506) In light of that, I think it might be reasonable to just rename them from `LICENSE.md` to `LICENSE` (maybe removing the `# License` lines also), which doesn't match the pattern that prettier is using to decide what should be formatted.

Also, we should enforce that the formatting is correct during the `lint` in `.github/workflows/tests.yml`. I think adding something like this should work:

```
- name: Check formatting
  run: npm run format && git diff --exit-code
```

**Comment by abdonrd** [comment]
@bicknellr done!

- `master` is merged
- Setup `ignore sync` (is what we use in `lit-html` mono-repo). https://github.com/webcomponents/polyfills/pull/405/commits/781c3d5bc3a0c33436ff10e8ccd1014846ebb699
- Add format step to CI. https://github.com/webcomponents/polyfills/pull/405/commits/421e1315f0e59d8fcb473b21cf27b1fe6b8f8912

**Comment by bicknellr** [comment]
Neat! I didn't know about `ignore-sync` before - seems like just what we needed here. Also, I pushed a few extra commits in a [local branch (`format`)](https://github.com/webcomponents/polyfills/commits/format), PTAL. If those seem ok to you, then I think this is ready to go.

**Comment by abdonrd** [comment]
Thanks @bicknellr! Now I've caught those commits!

Except this (https://github.com/abdonrd/polyfills/commit/3dfa021029c740a015bd99be6f4c216260f6baa3); I don't know why, but the tests breaks if we remove it.

**Comment by abdonrd** [comment]
@bicknellr Maybe I did something wrong, because everything is already working! 🎉

**Comment by bicknellr** [comment]
I ran into that same problem yesterday when I pulled up the tests locally. This snippet was running before `eventPropertyNamesForElement` was defined:

https://github.com/webcomponents/polyfills/blob/bed5d99d65490c32b2931c71380badc5d1387046/packages/shadydom/src/patches/Element.js#L127-L129

Normally this would mean that `eventPropertyNamesForElement` is part of an import cycle, but it comes from `patch-events.js` and that module and the ones below it have no cycles. (There's at least one cycle that `patches/Element.js` itself is part of, but that shouldn't affect `eventPropertyNamesForElement`.) Then I stumbled on this in the build output:

```
@webcomponents/shadydom: gulp-google-closure-compiler: src/patches/Element.js:128: WARNING - [JSC_NAME_DEFINED_LATE] eventPropertyNamesForElement$$module$src$patch_events.forEach defined before its owner. eventPropertyNamesForElement$$mod
ule$src$patch_events is defined at src/patch-events.js:757                                                                                                                                                                                    
@webcomponents/shadydom:   eventPropertyNamesForElement.forEach((property) => {                                                                                                                                                               
@webcomponents/shadydom:   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                               
@webcomponents/shadydom: src/patches/HTMLElement.js:33: WARNING - [JSC_NAME_DEFINED_LATE] eventPropertyNamesForHTMLElement$$module$src$patch_events.forEach defined before its owner. eventPropertyNamesForHTMLElement$$module$src$patch_event
s is defined at src/patch-events.js:761                                                                                                                                                                                                       
@webcomponents/shadydom:   eventPropertyNamesForHTMLElement.forEach((property) => {                                                                                                                                                           
@webcomponents/shadydom:   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                           
@webcomponents/shadydom: 0 error(s), 2 warning(s), 77.1% typed                                                                                                                                                                                
```

Closure has a [`dependency_mode`](https://github.com/google/closure-compiler/wiki/Flags-and-Options#dependency-management) flag that defaults to `NONE`, which apparently just includes all of the files in the order they're listed. So, my assumption is that we'd just been lucky regarding the import ordering until now and that changing that line somehow affected the order that the files were given to Closure, which sometimes broke the output.

Was your build ever broken when _both_ `entry_point` was specified and those import lines weren't being ignored by prettier? I wonder if we should explicitly set `dependency_mode` to `PRUNE`; maybe `entry_point` doesn't necessarily imply ordering.

**Comment by abdonrd** [comment]
I have issues with the tests locally:

Screenshot 2021-01-28 at 00 11 37

**Comment by bicknellr** [comment]
Oh, was that what was happening for you before when you mentioned the tests weren't working? Or were they previously running but with failures?

**Comment by abdonrd** [comment]
> Oh, was that what was happening for you before when you mentioned the tests weren't working? Or were they previously running but with failures?

Yes, I already had this problem before. It's an issue with wct and Java in my laptop.

**Review by bicknellr — APPROVED** [review]

**Comment by bicknellr** [comment]
Ok, in that case I'm going to go ahead and merge. Maybe I'll post another PR later to handle the `dependency_mode` thing, but I don't think it strictly needs to be addressed here. Thanks for this PR and your patience!

**Pull request merged**